File: my-folder/0094-binary-tree-inorder-traversal/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var traverse = function(root, ar) {
    if (root === null) return null

    traverse(root.left, ar)
    ar.push(root.val)
    traverse(root.right, ar)
    
}

var inorderTraversal = function(root) {
    let ar = []
    traverse(root, ar)
    return ar
};

File: my-folder/0094-binary-tree-inorder-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        # Morris Inorder Traversal
        ar = []
        while root:
            if not root.left:
                ar.append(root.val)
                root = root.right
            else:
                temp = root.left
                while temp.right and temp.right != root:
                    temp = temp.right
                
                if temp.right == root:
                    ar.append(root.val)
                    root = root.right
                else:
                    temp.right = root
                    root = root.left
        
        return ar
